frontend/catalyst/from_plxpr/decompose.py

```diff
@@ -0,0 +1,218 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A transform for the new MLIR-based Catalyst decomposition system.
+"""
+
+
+from __future__ import annotations
+
+import inspect
+from collections.abc import Callable
+from copy import copy
+from typing import get_type_hints
+
+import jax
+import pennylane as qml
+
+# GraphSolutionInterpreter:
+from pennylane.decomposition import DecompositionGraph
+from pennylane.measurements import MidMeasureMP
+from pennylane.wires import WiresLike
+
+from catalyst.jax_primitives import decomposition_rule
+
+COMPILER_OPERATIONS_NUM_WIRES = {
+    "CNOT": 2,
+    "ControlledPhaseShift": 2,
+    "CRot": 2,
+    "CRX": 2,
+    "CRY": 2,
+    "CRZ": 2,
+    "CSWAP": 3,
+    "CY": 2,
+    "CZ": 2,
+    "Hadamard": 1,
+    "Identity": 1,
+    "IsingXX": 2,
+    "IsingXY": 2,
+    "IsingYY": 2,
+    "IsingZZ": 2,
+    "SingleExcitation": 2,
+    "DoubleExcitation": 4,
+    "ISWAP": 2,
+    "PauliX": 1,
+    "PauliY": 1,
+    "PauliZ": 1,
+    "PhaseShift": 1,
+    "PSWAP": 2,
+    "Rot": 1,
+    "RX": 1,
+    "RY": 1,
+    "RZ": 1,
+    "S": 1,
+    "SWAP": 2,
+    "T": 1,
+    "Toffoli": 3,
+    "U1": 1,
+    "U2": 1,
+    "U3": 1,
+}
+
+
+def create_decomposition_rule(func: Callable, op_name: str, num_wires: int):
+    """Create a decomposition rule from a function."""
+
+    sig_func = inspect.signature(func)
+    type_hints = get_type_hints(func)
+
+    args = {}
+    for name in sig_func.parameters.keys():
+        typ = type_hints.get(name, None)
+
+        # Skip tailing kwargs in the rules
+        if name == "__":
+            continue
+
+        if typ is float or name in ("phi", "theta", "omega", "delta"):
```

Comment:
Why do we hard-code these names? I imagine that some decomposition rules might be:
```
def some_decomp(*params, wires, ...):
    qml.RX(params[0], wires=wires[0])
    ...
```
What happens then?

Reply:
These are mostly temporary, as we need to ensure that all decomp rules in PL include type hinting. We should rely on types than names.

Reply:
Let me know your thoughts on this @astralcai! I have a WIP branch in PL that's supposed to fix all these things.

Reply:
Yup! they should always provide type hints and then we can remove this weird condition

Reply:
Sorry I'm not sure I understand the impact on the user here 🤔

Reply:
@isaacdevlugt we need all decomposition rules to be defined with [type hints](https://peps.python.org/pep-0484/) for a seamless integration with Catalyst. Users need to always specify types of args and kwargs in their custom decomposition rules.